File: windows/YOLOTerm.Core/Layout/LayoutEngine.cs
namespace YOLOTerm.Core.Layout;

/// <summary>
/// Pure function: (pane set, preset, container size, drag deltas) → pane rects
/// No UI types. Fixture-tested behavior.
/// </summary>
public class LayoutEngine
{
    /// <summary>
    /// Calculate layout for given panes, preset, and container size
    /// </summary>
    public List<PaneRect> Calculate(
        List<string> paneIds,
        LayoutPreset preset,
        ContainerSize containerSize,
        List<DragDelta>? dragDeltas = null,
        string? zoomedPane = null)
    {
        dragDeltas ??= new List<DragDelta>();

        // If a pane is zoomed, it fills the entire container
        if (zoomedPane != null)
        {
            return new List<PaneRect>
            {
                new PaneRect(
                    id: zoomedPane,
                    x: 0,
                    y: 0,
                    width: containerSize.Width,
                    height: containerSize.Height)
            };
        }

        int count = paneIds.Count;
        if (count == 0) return new List<PaneRect>();

        // Single pane always fills container
        if (count == 1)
        {
            return new List<PaneRect>
            {
                new PaneRect(
                    id: paneIds[0],
                    x: 0,
                    y: 0,
                    width: containerSize.Width,
                    height: containerSize.Height)
            };
        }

        // Apply preset
        var actualPreset = preset == LayoutPreset.Auto ? AutoPreset(count) : preset;

        return actualPreset switch
        {
            LayoutPreset.Single => LayoutSingle(paneIds, containerSize),
            LayoutPreset.Columns => LayoutColumns(paneIds, containerSize),
            LayoutPreset.Rows => LayoutRows(paneIds, containerSize),
            LayoutPreset.Grid => LayoutGrid(paneIds, containerSize),
            LayoutPreset.MainLeft => LayoutMainLeft(paneIds, containerSize),
            LayoutPreset.MainRight => LayoutMainRight(paneIds, containerSize),
            LayoutPreset.Auto => LayoutAuto(paneIds, containerSize),
            _ => throw new ArgumentException($"Unknown preset: {actualPreset}")
        };
    }

    /// <summary>
    /// Equalize all panes (reset custom sizing)
    /// </summary>
    public List<PaneRect> Equalize(
        List<string> paneIds,
        LayoutPreset preset,
        ContainerSize containerSize)
    {
        // Equalize just recalculates without drag deltas
        return Calculate(paneIds, preset, containerSize, new List<DragDelta>());
    }

    // MARK: - Private Layout Algorithms

    private LayoutPreset AutoPreset(int count)
    {
        return count switch
        {
            1 => LayoutPreset.Single,
            2 => LayoutPreset.Columns,
            3 => LayoutPreset.MainLeft,
            4 => LayoutPreset.Grid,
            _ => LayoutPreset.Grid // 5+ panes: use grid
        };
    }

    private List<PaneRect> LayoutSingle(List<string> paneIds, ContainerSize containerSize)
    {
        if (paneIds.Count == 0) return new List<PaneRect>();

        return new List<PaneRect>
        {
            new PaneRect(
                id: paneIds[0],
                x: 0,
                y: 0,
                width: containerSize.Width,
                height: containerSize.Height)
        };
    }

    private List<PaneRect> LayoutColumns(List<string> paneIds, ContainerSize containerSize)
    {
        int count = paneIds.Count;
        double colWidth = containerSize.Width / count;

        var rects = new List<PaneRect>();
        for (int i = 0; i < count; i++)
        {
            rects.Add(new PaneRect(
                id: paneIds[i],
                x: i * colWidth,
                y: 0,
                width: colWidth,
                height: containerSize.Height));
        }
        return rects;
    }

    private List<PaneRect> LayoutRows(List<string> paneIds, ContainerSize containerSize)
    {
        int count = paneIds.Count;
        double rowHeight = containerSize.Height / count;

        var rects = new List<PaneRect>();
        for (int i = 0; i < count; i++)
        {
            rects.Add(new PaneRect(
                id: paneIds[i],
                x: 0,
                y: i * rowHeight,
                width: containerSize.Width,
                height: rowHeight));
        }
        return rects;
    }

    private List<PaneRect> LayoutGrid(List<string> paneIds, ContainerSize containerSize)
    {
        int count = paneIds.Count;

        // Calculate grid dimensions
        int cols = (int)Math.Ceiling(Math.Sqrt(count));
        int rows = (int)Math.Ceiling((double)count / cols);

        double colWidth = containerSize.Width / cols;
        double rowHeight = containerSize.Height / rows;

        var rects = new List<PaneRect>();
        for (int i = 0; i < count; i++)
        {
            int col = i % cols;
            int row = i / cols;

            rects.Add(new PaneRect(
                id: paneIds[i],
                x: col * colWidth,
                y: row * rowHeight,
                width: colWidth,
                height: rowHeight));
        }
        return rects;
    }

    private List<PaneRect> LayoutMainLeft(List<string> paneIds, ContainerSize containerSize)
    {
        if (paneIds.Count < 2)
        {
            return LayoutSingle(paneIds, containerSize);
        }

        // Main pane on left takes 50%, rest stacked on right
        double mainWidth = containerSize.Width * 0.5;
        double sideWidth = containerSize.Width * 0.5;

        var rects = new List<PaneRect>();

        // Main pane
        rects.Add(new PaneRect(
            id: paneIds[0],
            x: 0,
            y: 0,
            width: mainWidth,
            height: containerSize.Height));

        // Side panes stacked vertically
        int sideCount = paneIds.Count - 1;
        double sideHeight = containerSize.Height / sideCount;

        for (int i = 0; i < sideCount; i++)
        {
            rects.Add(new PaneRect(
                id: paneIds[i + 1],
                x: mainWidth,
                y: i * sideHeight,
                width: sideWidth,
                height: sideHeight));
        }

        return rects;
    }

    private List<PaneRect> LayoutMainRight(List<string> paneIds, ContainerSize containerSize)
    {
        if (paneIds.Count < 2)
        {
            return LayoutSingle(paneIds, containerSize);
        }

        // Main pane on right takes 50%, rest stacked on left
        double sideWidth = containerSize.Width * 0.5;
        double mainWidth = containerSize.Width * 0.5;

        var rects = new List<PaneRect>();

        // Side panes stacked vertically
        int sideCount = paneIds.Count - 1;
        double sideHeight = containerSize.Height / sideCount;

        for (int i = 0; i < sideCount; i++)
        {
            rects.Add(new PaneRect(
                id: paneIds[i + 1],
                x: 0,
                y: i * sideHeight,
                width: sideWidth,
                height: sideHeight));
        }

        // Main pane
        rects.Add(new PaneRect(
            id: paneIds[0],
            x: sideWidth,
            y: 0,
            width: mainWidth,
            height: containerSize.Height));

        return rects;
    }

    private List<PaneRect> LayoutAuto(List<string> paneIds, ContainerSize containerSize)
    {
        var actualPreset = AutoPreset(paneIds.Count);
        return Calculate(paneIds, actualPreset, containerSize);
    }
}

File: macos/Sources/YOLOTermKit/Layout/LayoutEngine.swift
import Foundation

// MARK: - Layout Types (Pure Swift, No AppKit)

/// A rectangle representing a pane's position and size
public struct PaneRect: Equatable, Sendable {
    public let id: String
    public let x: Double
    public let y: Double
    public let width: Double
    public let height: Double
    
    public init(id: String, x: Double, y: Double, width: Double, height: Double) {
        self.id = id
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

/// Container size
public struct ContainerSize: Equatable, Sendable {
    public let width: Double
    public let height: Double
    
    public init(width: Double, height: Double) {
        self.width = width
        self.height = height
    }
}

/// Layout preset types
public enum LayoutPreset: String, Sendable {
    case auto
    case single
    case columns
    case rows
    case grid
    case mainLeft = "main-left"
    case mainRight = "main-right"
}

/// Drag delta for border adjustments
public struct DragDelta: Sendable {
    public enum Orientation: Sendable {
        case horizontal
        case vertical
    }
    
    public let orientation: Orientation
    public let position: Double // normalized 0-1
    public let delta: Double // in pixels
    
    public init(orientation: Orientation, position: Double, delta: Double) {
        self.orientation = orientation
        self.position = position
        self.delta = delta
    }
}

// MARK: - Layout Engine

/// Pure function: (pane set, preset, container size, drag deltas) → pane rects
/// No UI types. Fixture-tested behavior.
public struct LayoutEngine: Sendable {
    
    public init() {}
    
    /// Calculate layout for given panes, preset, and container size
    public func calculate(
        paneIds: [String],
        preset: LayoutPreset,
        containerSize: ContainerSize,
        dragDeltas: [DragDelta] = [],
        zoomedPane: String? = nil
    ) -> [PaneRect] {
        // If a pane is zoomed, it fills the entire container
        if let zoomedId = zoomedPane {
            return [PaneRect(
                id: zoomedId,
                x: 0,
                y: 0,
                width: containerSize.width,
                height: containerSize.height
            )]
        }
        
        let count = paneIds.count
        guard count > 0 else { return [] }
        
        // Single pane always fills container
        if count == 1 {
            return [PaneRect(
                id: paneIds[0],
                x: 0,
                y: 0,
                width: containerSize.width,
                height: containerSize.height
            )]
        }
        
        // Apply preset
        let actualPreset = preset == .auto ? autoPreset(for: count) : preset
        
        switch actualPreset {
        case .single:
            return layoutSingle(paneIds: paneIds, containerSize: containerSize)
        case .columns:
            return layoutColumns(paneIds: paneIds, containerSize: containerSize)
        case .rows:
            return layoutRows(paneIds: paneIds, containerSize: containerSize)
        case .grid:
            return layoutGrid(paneIds: paneIds, containerSize: containerSize)
        case .mainLeft:
            return layoutMainLeft(paneIds: paneIds, containerSize: containerSize)
        case .mainRight:
            return layoutMainRight(paneIds: paneIds, containerSize: containerSize)
        case .auto:
            // Shouldn't reach here as auto is resolved above
            return layoutAuto(paneIds: paneIds, containerSize: containerSize)
        }
    }
    
    /// Equalize all panes (reset custom sizing)
    public func equalize(
        paneIds: [String],
        preset: LayoutPreset,
        containerSize: ContainerSize
    ) -> [PaneRect] {
        // Equalize just recalculates without drag deltas
        return calculate(
            paneIds: paneIds,
            preset: preset,
            containerSize: containerSize,
            dragDeltas: []
        )
    }
    
    // MARK: - Private Layout Algorithms
    
    private func autoPreset(for count: Int) -> LayoutPreset {
        switch count {
        case 1:
            return .single
        case 2:
            return .columns
        case 3:
            return .mainLeft
        case 4:
            return .grid
        default:
            // 5+ panes: use grid
            return .grid
        }
    }
    
    private func layoutSingle(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        guard let firstId = paneIds.first else { return [] }
        return [PaneRect(
            id: firstId,
            x: 0,
            y: 0,
            width: containerSize.width,
            height: containerSize.height
        )]
    }
    
    private func layoutColumns(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        let count = paneIds.count
        let colWidth = containerSize.width / Double(count)
        
        return paneIds.enumerated().map { index, id in
            PaneRect(
                id: id,
                x: Double(index) * colWidth,
                y: 0,
                width: colWidth,
                height: containerSize.height
            )
        }
    }
    
    private func layoutRows(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        let count = paneIds.count
        let rowHeight = containerSize.height / Double(count)
        
        return paneIds.enumerated().map { index, id in
            PaneRect(
                id: id,
                x: 0,
                y: Double(index) * rowHeight,
                width: containerSize.width,
                height: rowHeight
            )
        }
    }
    
    private func layoutGrid(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        let count = paneIds.count
        
        // Calculate grid dimensions
        let cols = Int(ceil(sqrt(Double(count))))
        let rows = Int(ceil(Double(count) / Double(cols)))
        
        let colWidth = containerSize.width / Double(cols)
        let rowHeight = containerSize.height / Double(rows)
        
        return paneIds.enumerated().map { index, id in
            let col = index % cols
            let row = index / cols
            
            return PaneRect(
                id: id,
                x: Double(col) * colWidth,
                y: Double(row) * rowHeight,
                width: colWidth,
                height: rowHeight
            )
        }
    }
    
    private func layoutMainLeft(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        guard paneIds.count >= 2 else {
            return layoutSingle(paneIds: paneIds, containerSize: containerSize)
        }
        
        // Main pane on left takes 50%, rest stacked on right
        let mainWidth = containerSize.width * 0.5
        let sideWidth = containerSize.width * 0.5
        
        let mainPane = PaneRect(
            id: paneIds[0],
            x: 0,
            y: 0,
            width: mainWidth,
            height: containerSize.height
        )
        
        let sideCount = paneIds.count - 1
        let sideHeight = containerSize.height / Double(sideCount)
        
        let sidePanes = paneIds.dropFirst().enumerated().map { index, id in
            PaneRect(
                id: id,
                x: mainWidth,
                y: Double(index) * sideHeight,
                width: sideWidth,
                height: sideHeight
            )
        }
        
        return [mainPane] + sidePanes
    }
    
    private func layoutMainRight(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        guard paneIds.count >= 2 else {
            return layoutSingle(paneIds: paneIds, containerSize: containerSize)
        }
        
        // Main pane on right takes 50%, rest stacked on left
        let sideWidth = containerSize.width * 0.5
        let mainWidth = containerSize.width * 0.5
        
        let sideCount = paneIds.count - 1
        let sideHeight = containerSize.height / Double(sideCount)
        
        let sidePanes = paneIds.dropFirst().enumerated().map { index, id in
            PaneRect(
                id: id,
                x: 0,
                y: Double(index) * sideHeight,
                width: sideWidth,
                height: sideHeight
            )
        }
        
        let mainPane = PaneRect(
            id: paneIds[0],
            x: sideWidth,
            y: 0,
            width: mainWidth,
            height: containerSize.height
        )
        
        return sidePanes + [mainPane]
    }
    
    private func layoutAuto(paneIds: [String], containerSize: ContainerSize) -> [PaneRect] {
        let actualPreset = autoPreset(for: paneIds.count)
        return calculate(
            paneIds: paneIds,
            preset: actualPreset,
            containerSize: containerSize
        )
    }
}
